src/jpgimage.cpp

```diff
@@ -350,41 +363,48 @@ namespace Exiv2 {
         }
         clearMetadata();
         int search = 6 ; // Exif, ICC, XMP, Comment, IPTC, SOF
-        const long bufMinSize = 36;
-        long bufRead = 0;
-        DataBuf buf(bufMinSize);
         Blob psBlob;
         bool foundCompletePsData = false;
         bool foundExifData = false;
         bool foundXmpData = false;
         bool foundIccData = false;
 
+        // which markers have a length field?
+        // TODO: move this to a utility function
+        bool mHasLength[256];
+        for (int i = 0; i < 256; i++)
+            mHasLength[i] = (i >= sof0_ && i <= sof15_) || (i >= app0_ && i <= (app0_ | 0x0F)) ||
+                            (i == dht_ || i == dqt_ || i == dri_ || i == com_ || i == sos_);
+
         // Read section marker
         int marker = advanceToMarker();
         if (marker < 0) throw Error(kerNotAJpeg);
 
         while (marker != sos_ && marker != eoi_ && search > 0) {
-            // Read size and signature (ok if this hits EOF)
-            std::memset(buf.pData_, 0x0, buf.size_);
-            bufRead = io_->read(buf.pData_, bufMinSize);
-            if (io_->error()) throw Error(kerFailedToReadImageData);
-            if (bufRead < 2) throw Error(kerNotAJpeg);
-            uint16_t size = getUShort(buf.pData_, bigEndian);
+            // 2-byte buffer for reading the size.
+            byte sizebuf[2];
+            uint16_t size = 0;
+            if (mHasLength[marker]) {
+                readOrThrow(*io_, sizebuf, 2, kerFailedToReadImageData);
+                size = getUShort(sizebuf, bigEndian);
+                // `size` is the size of the segment, including the 2-byte size field
+                // that we just read.
+                enforce(size >= 2, kerFailedToReadImageData);
+            }
+
+            // Read the rest of the segment.
+            DataBuf buf(size);
+            if (size > 0) {
+                assert(size >= 2); // enforced above
+                readOrThrow(*io_, buf.pData_ + 2, size - 2, kerFailedToReadImageData);
+                memcpy(buf.pData_, sizebuf, 2);
+            }
 
             if (   !foundExifData
-                && marker == app1_ && memcmp(buf.pData_ + 2, exifId_, 6) == 0) {
-                if (size < 8) {
-                    rc = 1;
-                    break;
-                }
-                // Seek to beginning and read the Exif data
-                io_->seek(8 - bufRead, BasicIo::cur);
-                DataBuf rawExif(size - 8);
-                io_->read(rawExif.pData_, rawExif.size_);
-                if (io_->error() || io_->eof()) throw Error(kerFailedToReadImageData);
-                ByteOrder bo = ExifParser::decode(exifData_, rawExif.pData_, rawExif.size_);
+                && marker == app1_ && size >= 8 && memcmp(buf.pData_ + 2, exifId_, 6) == 0) {
```

Comment:
Looking at `size>=8` and `size>=31` below, do you know where these magic values come from? Are they known constants ?

If so I think it might be worth to add a comment explaining their origin.

Reply:
It's to prevent an out-of-bounds read in the `memcmp`. I'll add a comment.